daffodil-lib/src/main/scala/edu/illinois/ncsa/daffodil/util/DecimalUtils.scala

```diff
@@ -0,0 +1,342 @@
+      case BinaryNumberCheckPolicy.Lax => Map("positive" -> valid_positives, "negative" -> valid_negatives, "unsigned" -> valid_unsigneds, "zero_sign" -> valid_zero_signs)
+    }
+
+    hexCodes
+  }
+
+  def packedToBigInteger(num: Array[Byte], signCodes: String, policy: BinaryNumberCheckPolicy): JBigInteger = {
+    val hexCodes: Map[String, List[Int]] = signCodesToHex(signCodes, policy)
+    val numDigits: Int = num.size * 2  // 2 digits stored per byte
+    val outputData: Array[Char] = new Array[Char](numDigits-1)
+    var outputPos: Int = 0
+    var offset: Int = 0
+    var nibble: Int = 0
+    var negative: Boolean = false
+
+    // Parse and validate the last (sign) bit
+    nibble = (num(offset +  num.size - 1) & 0x0F)
+    if (hexCodes("negative").contains(nibble)) {
+      negative = true
+    } else if (!hexCodes("positive").contains(nibble)) {
+      throw new NumberFormatException("Invalid low nibble")
+    }
+
+    while (outputPos < outputData.size - 1) {
+      // Parse high nibble
+      nibble = (num(offset) & 0xFF) >>> 4
+      if (nibble > 0x09) {
+        throw new NumberFormatException("Invalid high nibble")
+      }
+
+      outputData(outputPos) = (nibble | 0x0030).toChar
+      outputPos = outputPos + 1
+
+      // Parse low nibble
+      nibble = (num(offset) & 0x0F)
+      offset = offset + 1
+      if (nibble > 0x09) {
+        throw new NumberFormatException("Invalid low nibble")
+      }
+
+      outputData(outputPos) = (nibble | 0x0030).toChar
+      outputPos = outputPos + 1
+    }
+
+    // Parse last digit
+    nibble = (num(offset) & 0xFF) >>> 4
+    if (nibble > 0x09) {
+      throw new NumberFormatException("Invalid high nibble")
+    }
+
+    outputData(outputPos) = (nibble | 0x0030).toChar
+
+    if (negative)
+      new JBigInteger(new String(outputData)).negate()
+    else
+      new JBigInteger(new String(outputData))
+
+  }
+
+  def packedToBigDecimal(num: Array[Byte], scale: Int, signCodes: String, policy: BinaryNumberCheckPolicy): JBigDecimal = {
+    return new JBigDecimal(packedToBigInteger(num, signCodes, policy), scale)
+  }
+
+  def packedFromBigInteger(num: String, nBits: Int, signCodes:String, policy: BinaryNumberCheckPolicy): Array[Byte] = {
+    val negative: Boolean = (num.charAt(0) == '-')
+    // Discard any sign symbols or decimal points from the string (decimal point will be maintained in binaryVirtualDecimalPoint)
+    val inChars: Array[Char] = num.replaceAll("[^0-9]", "").replaceFirst("^0+(?!$)", "").toCharArray
+    val hexCodes: Map[String, List[Int]] = signCodesToHex(signCodes, policy)
+    var offset: Int = 0
+    var inPos: Int = 0
+    var firstNibble: Int = 0
+    var secondNibble: Int = 0
+    // Don't count the '-' symbol when determing the number of digits
+    val numDigits: Int = inChars.length
+    val outArray: Array[Byte] = new Array[Byte](nBits/8)
+    val leadingZeros: Int = if(numDigits % 2 == 0) (nBits/4 - numDigits - 1) else (nBits/4 - numDigits)
+
```

Comment:
Some comment above about types, not necesary. Also, syntax standard-wise, check your parens after keywords like if(foo) and while(foo). There should be a space after the keyword.